### PR TITLE
Fixed issue when escape key is pressed to close prompt

### PR DESCRIPTION
**Pull request by konarshankar07**
### Related Pull Requests
https://github.com/magento/partners-magento2b2b/pull/49

### Description (*)

This PR will fix the issue with prompt. Currently, If escape key is pressed then prompt is not closing but instead of closing it try to submit the prompt
### Fixed Issues (if relevant)

1. magento/adobe-stock-integration#572: Pressing escape key confirms the save/license confirmation popup

### Manual testing scenarios (*)

1. Login to admin panel
2. Open Magento Media Gallery (i.e. go to Cms -> Pages, edit the page and insert image)
3. Click "New Folder" button to open prompt
4. Enter `Magento` in the textbox to create folder
4. Press "Escape" key on the keyboard

### Actual result
Trying to create new folder with `magento`

### Expected result
Popup should be closed and new folder should not be created

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-6284](https://jira.corp.magento.com/browse/ENGCOM-6284) has been created to process this Pull Request
:eight_spoked_asterisk: @sivaschenko, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Delta** [comment]
Hi @konarshankar07 During testing PR I found this issue: **"New Folder Name"** only hides behind **"Media Gallery"** popup if press "Escape" button but it is not closed
#25349PR

Could you take a look?

**Comment by konarshankar07** [comment]
Hello @engcom-Delta ,
Thanks for raising this issue. I have fixed the above mention issue. Please test and let me know your feedback

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-6284](https://jira.corp.magento.com/browse/ENGCOM-6284) has been created to process this Pull Request
:eight_spoked_asterisk: @sivaschenko, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Review by VladimirZaets — CHANGES_REQUESTED** [review]
Hi @konarshankar07. This problem is caused because this fix was merged https://github.com/magento/magento2/pull/16477. We should revert it and investigate the first problem that described in the issue that I linked

**Comment by konarshankar07** [comment]
Hello @VladimirZaets ,
Can you please add the link for the issue so that I can work on it? or Can you please confirm the issue https://github.com/magento/magento2/issues/14593 that you are talking about it?
Thanks

**Comment by VladimirZaets** [comment]
@konarshankar07, yes, correct. We should invistigate the first https://github.com/magento/magento2/issues/14593 issue

**Comment by konarshankar07** [comment]
Hello @VladimirZaets ,
I tried to reproduce the error which is mentioned in the task but I'm not able to reproduce it. Here is the video https://www.loom.com/share/be625ed555ac4cd58a2ad543d1ff67dd. Please let me know if I'm missing something here.
Thanks

**Comment by konarshankar07** [comment]
Hello @VladimirZaets ,
I have reverted the changes for https://github.com/magento/magento2/pull/16477 PR but I'm not able to understand the issue here with my PR. I opened this PR for `prompt` component but the mention PR in above component have issue in `modal` component. I know `prompt` component extends the `modal` component but I'm not able to understand the issue with the mention PR. If explain in more detail then it would be helpful to fix both the issues.
Thanks

**Comment by VladimirZaets** [comment]
@konarshankar07 this issue with the prom caused by PR that I mention. If that PR will be reverted this issue will not be reproducing

**Comment by konarshankar07** [comment]
Thanks @VladimirZaets , I'm now able to reproduce the issue and will try to fix it

File: app/code/Magento/Checkout/view/frontend/web/js/sidebar.js [hunk 1]
```diff
@@ -98,7 +98,9 @@ define([
 
                         /** @inheritdoc */
                         always: function (e) {
-                            e.stopImmediatePropagation();
+                            if (e) {
```
Comment (VladimirZaets) [inline app/code/Magento/Checkout/view/frontend/web/js/sidebar.js]: Do we really need to check is event exists? It's framework lifecycle methods and the event should be always presented

**Review by VladimirZaets — CHANGES_REQUESTED** [review]
Hi, @konarshankar07. Please take a look at my comment

File: app/code/Magento/Checkout/view/frontend/web/js/sidebar.js (hunk 1 above)
Comment (konarshankar07) [reply app/code/Magento/Checkout/view/frontend/web/js/sidebar.js]: yes, we need to check is event exist cause as you see we are not passing any event in the `closeModal()` function and if we pass the event function then prompt will not close if we press the escape key cause in `prompt` component again we are not passing event in the `closeModal()`

Comment (sivaschenko) [reply app/code/Magento/Checkout/view/frontend/web/js/sidebar.js]: @konarshankar07 @VladimirZaets should we just remove the `always` function?

Comment (konarshankar07) [reply app/code/Magento/Checkout/view/frontend/web/js/sidebar.js]: @sivaschenko ... I don't think we can remove `always` function cause it depends on [here](https://github.com/magento/magento2/blob/c7ca62eaa2d5e6a8978dc9eeb00d14625c3ef893/app/code/Magento/Ui/view/base/web/js/modal/prompt.js#L169) but still not sure

File: app/code/Magento/Ui/view/base/web/js/modal/modal.js [hunk 2]
```diff
@@ -104,7 +104,6 @@ define([
                 /**
                  * Escape key press handler,
                  * close modal window
-                 * @param {Object} event - event
```
Comment (VladimirZaets) [inline app/code/Magento/Ui/view/base/web/js/modal/modal.js]: Why you remove param description if this param still passing to `escapeKey` method?

Comment (konarshankar07) [reply app/code/Magento/Ui/view/base/web/js/modal/modal.js]: Sorry, I have missed the parameter description but now I have added

**Review by krzksz — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @krzksz, thank you for the review.
[ENGCOM-6284](https://jira.corp.magento.com/browse/ENGCOM-6284) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
Hi @konarshankar07 I am able to reproduce issue described by this PR with latest changes on PR branch.
Could you take a look?

**Comment by konarshankar07** [comment]
@magento run all tests

**Comment by sivaschenko** [comment]
@magento run all tests

File: app/code/Magento/Ui/view/base/web/js/modal/modal.js [hunk 3]
```diff
@@ -181,7 +181,7 @@ define([
             var key = keyCodes[event.keyCode];
 
             if (this.options.keyEventHandlers.hasOwnProperty(key)) {
-                this.options.keyEventHandlers[key].apply(this, arguments);
+                this.options.keyEventHandlers[key].apply(this, arguments, event);
```
Comment (omiroshnichenko) [inline app/code/Magento/Ui/view/base/web/js/modal/modal.js]: Please, revert these changes. Apply function doesn't accept the third argument.

File: app/code/Magento/Ui/view/base/web/js/modal/prompt.js [hunk 4]
```diff
@@ -154,8 +154,7 @@ define([
          */
         closeModal: function (result) {
```
Comment (omiroshnichenko) [inline app/code/Magento/Ui/view/base/web/js/modal/prompt.js]: Modify [these](https://github.com/magento/magento2/pull/25349/files#diff-9458ab0c3f9705f0f9bdb6fc1b15907dR47-R68) functions  to accept the event as an argument and use it in closeModal function call as a first argument(same as in [confirm.js](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Ui/view/base/web/js/modal/confirm.js#L40-L60)). After that, start accepting in closeModal event as first argument and result as second([example](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Ui/view/base/web/js/modal/confirm.js#L89)).

**Review by omiroshnichenko — CHANGES_REQUESTED** [review]
See comments on changed files

File: app/code/Magento/Ui/view/base/web/js/modal/prompt.js [hunk 5]
```diff
@@ -61,8 +61,8 @@ define([
                 /**
                  * Click handler.
                  */
-                click: function () {
-                    this.closeModal(true);
+                click: function (event) {
+                    this.closeModal(event);
```
Comment (sivaschenko) [inline app/code/Magento/Ui/view/base/web/js/modal/prompt.js]: Should the result be passed as well?
```suggestion
                    this.closeModal(event, true);
```

File: app/code/Magento/Ui/view/base/web/js/modal/prompt.js [hunk 6]
```diff
@@ -152,10 +152,9 @@ define([
         /**
          * Close modal window
          */
-        closeModal: function (result) {
+        closeModal: function (event, result) {
             var value;
-
-            if (result) {
+            if (result && !(result instanceof $.Event)) {
```
Comment (sivaschenko) [inline app/code/Magento/Ui/view/base/web/js/modal/prompt.js]: What is the purpose of the `instanceof` check?

File: app/code/Magento/Ui/view/base/web/js/modal/prompt.js [hunk 7]
```diff
@@ -152,10 +152,9 @@ define([
         /**
          * Close modal window
          */
-        closeModal: function (result) {
+        closeModal: function (event, result) {
```
Comment (sivaschenko) [inline app/code/Magento/Ui/view/base/web/js/modal/prompt.js:155]: Should the `event` be used somehow in the method body?

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for update @konarshankar07 ! Please see my review comments

@omiroshnichenko can you please review the changes as well?

File: app/code/Magento/Ui/view/base/web/js/modal/prompt.js (hunk 6 above)
Comment (konarshankar07) [reply app/code/Magento/Ui/view/base/web/js/modal/prompt.js]: Actually, Initially we are passing only event so when this if condition is executed then result is taking as true so I've additional condition added that result should be instance of the event but I think we don't need now cause I have added condition if result is not passed then result will take as false

File: app/code/Magento/Ui/view/base/web/js/modal/prompt.js (hunk 5 above)
Comment (konarshankar07) [reply app/code/Magento/Ui/view/base/web/js/modal/prompt.js]: I have added condition if result is not passed then result will take as false

File: app/code/Magento/Ui/view/base/web/js/modal/prompt.js (hunk 7 above)
Comment (konarshankar07) [reply app/code/Magento/Ui/view/base/web/js/modal/prompt.js:155]: Yes, You are right and event is used in the method body

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-6284](https://jira.corp.magento.com/browse/ENGCOM-6284) has been created to process this Pull Request

**Comment by sivaschenko** [comment]
@magento run Static tests

**Comment by sivaschenko** [comment]
@magento run Static Tests

**Comment by sivaschenko** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
Related PR with B2B fixes https://github.com/magento/partners-magento2b2b/pull/49

**Review by slavvka — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @slavvka, thank you for the review.
[ENGCOM-6284](https://jira.corp.magento.com/browse/ENGCOM-6284) has been created to process this Pull Request

**Comment by naydav** [comment]
@konarshankar07 @slavvka @sivaschenko
Hello guys,

Unfortunately, we need to revert these changes. In the scope of this PR we changed the base  UI components ("modal", "modal/promt"). Thus it can lead to a huge impact on the current existed implementations based on this component.

We can accept some backward-incompatible changes but in this case, it is unreasonably.
We need to reimplement the fix in another way.

Thanks

**Comment by davemacaulay** [comment]
@naydav & others, this PR broke Page Builder's usage of the modal and prompt substantially. I agree this BiC change is maybe too severe as it has potential to break any consumer of these libraries.

**Comment by konarshankar07** [comment]
Hello @naydav ,
Sorry for the trouble caused by this PR but can you please explain in detail? So I can fix this issue in another PR. Honestly, I don't understand how this PR caused the issue after passing all the tests.
Thanks

**Comment by slavvka** [comment]
@konarshankar07 You changed the signature of methods in BIC way. Please look the guide https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/. All that rules apply to all code (both PHP and JS). But unfortunately our SVC doesn't check that for JS files. As you could see that already made your fix incompatible with B2B (that required changes there as well). Also we have some extensions that aren't used to run tests - so after running the builds with them we discovered the issue with PageBuilder. As therefore this could be a huge issue for all other marketplace extensions that could use those methods. That is why we ask you to reimplement the fix in BC way according to the guide and create a new PR. This PR has been reverted.

**Comment by sivaschenko** [comment]
@naydav @slavvka @davemacaulay this pull request is fixing a major issue of popups functionality as well as eliminating the interface inconsistency. I'd recommend reviewing the possibility to include this fix in minor release (2.4)

**Comment by naydav** [comment]
@sivaschenko 
We can make fix in compatible way and deliver it in 2.4
